LocalPV e2e test with two pods writing/reading to a shared file on one PVC

ianchakeres opened the pull request:
**What this PR does / why we need it**:

This PR adds an e2e test where one pod writes to a localPV and another reads from it. This e2e test is for the PersistentLocalVolumes alpha feature.

**Which issue this PR fixes**: fixes https://github.com/kubernetes/kubernetes/issues/48832

**Special notes for your reviewer**:

You need to setup the PersistentLocalVolumes feature-gate and NODE_LOCAL_SSDS environment variable.

For example:

```
KUBE_FEATURE_GATES="PersistentLocalVolumes=true" NODE_LOCAL_SSDS=1 go run hack/e2e.go -- -v --up

go run hack/e2e.go -- -v --test --test_args="--ginkgo.focus=\[Feature:LocalPersistentVolumes\]"

```
**Release note**:

```release-note
NONE
```

ianchakeres committed:
Added localPV e2e tests with two pods and refactored existing tests

ianchakeres commented:
/assign @msau42 - Let me know what you think of this e2e test.

ianchakeres commented:
@k8s-bot ok to test

ianchakeres commented:
/sig storage

msau42 commented on test/e2e/storage/persistent_volumes-local.go:
```diff
@@ -132,6 +132,42 @@ var _ = framework.KubeDescribe("[Volume] PersistentVolumes-local [Feature:LocalP
+
+			By("Creating a pod to write to the PV")
+			testFilePath := filepath.Join("/mnt/volume1", testFile)
+			cmd := fmt.Sprintf("echo %s > %s; cat %s", testVol.hostDir, testFilePath, testFilePath)
```
Can you put the read and write commands into wrappers, since many of the tests are doing the same thing?

msau42 commented on test/e2e/storage/persistent_volumes-local.go:
```diff
@@ -132,8 +134,171 @@ var _ = framework.KubeDescribe("[Volume] PersistentVolumes-local [Feature:LocalP
+		})
+	})
+
+	Context("when two pods request one prebound PVC one after other", func() {
```
I see the serially test above is using `TestContainerOutput`, which runs the passed in command to completion, whereas this test is creating a long running pod, execing into it to run the commands, and then destroying the pod.  

I'm curious to why there is difference in behavior (above is not seeing the unmount bug, but this one is)?  Is it that the methods for checking pod termination state are different?  Theoretically, the unmount bug should exist regardless of if the pod is long running or run-once.

msau42 commented on test/e2e/storage/persistent_volumes-local.go:
```diff
@@ -132,8 +134,171 @@ var _ = framework.KubeDescribe("[Volume] PersistentVolumes-local [Feature:LocalP
 
+// podNode wraps RunKubectl to get node where pod is running
+func podNodeName(pod *v1.Pod) string {
+	return strings.TrimSpace(framework.RunKubectlOrDie("get", "pod", pod.Name, fmt.Sprintf("--namespace=%s", pod.Namespace), "--no-headers", "-o=custom-columns=NODE:.spec.nodeName"))
```
You can use the go client APIs to get the information instead of parsing text output.  `conflig.client.CoreV1().Pods(config.ns).Get(podName)`

msau42 commented on test/e2e/storage/persistent_volumes-local.go:
```diff
@@ -132,8 +134,171 @@ var _ = framework.KubeDescribe("[Volume] PersistentVolumes-local [Feature:LocalP
+			Expect(pod1NodeName).To(Equal(node.Name))
+
+			//delete pod1
+			framework.DeletePodOrFail(config.client, config.ns, pod1.Name)
```
Also include a `By()` when you delete the pod, then we can distinguish what is parallel and what is sequential.

msau42 commented on test/e2e/storage/persistent_volumes-local.go:
```diff
@@ -132,8 +134,171 @@ var _ = framework.KubeDescribe("[Volume] PersistentVolumes-local [Feature:LocalP
+			//delete pod1
+			framework.DeletePodOrFail(config.client, config.ns, pod1.Name)
+
+			By("Creating a pod to read from the PV")
```
Maybe call this the "second pod"

msau42 commented on test/e2e/storage/persistent_volumes-local.go:
```diff
@@ -132,8 +134,171 @@ var _ = framework.KubeDescribe("[Volume] PersistentVolumes-local [Feature:LocalP
+			pod1, pod1Err := config.client.CoreV1().Pods(config.ns).Create(podSpec1)
+			Expect(pod1Err).NotTo(HaveOccurred())
+
+			framework.ExpectNoError(framework.WaitForPodRunningInNamespace(config.client, pod1))
```
Would you be able to use `framework.CreatePod()`, which creates the pod, waits for it to be running, and refreshes the pod status.

msau42 commented on test/e2e/storage/persistent_volumes-local.go:
```diff
@@ -132,8 +134,171 @@ var _ = framework.KubeDescribe("[Volume] PersistentVolumes-local [Feature:LocalP
+			framework.Logf("Pod2 %q created on Node %q", pod2.Name, pod2NodeName)
+			Expect(pod2NodeName).To(Equal(node.Name))
+
+			_, writeErr := podExec(pod1, writeCmd)
```
It would be good to put the read/write and exec into wrappers since most of these tests are utilizing them.

ianchakeres commented:
@msau42 @saad-ali @jeffvance @copejon @erinboyd - PTAL. I refactored all the tests to use common methods and added a few more local storage e2e tests. Note: for these tests to run successfully you need this bugfix: https://github.com/kubernetes/kubernetes/pull/48402

jeffvance commented on test/e2e/storage/persistent_volumes-local.go:
```diff
@@ -105,43 +110,286 @@ var _ = framework.KubeDescribe("[Volume] PersistentVolumes-local [Feature:LocalP
+		})
+
+		AfterEach(func() {
+			cleanupLocalVolume(config, testVol)
```
Only Logf is called in cases of errors. Do you want to call Failf or Expect in this func?

ianchakeres replied:
@msau42 - I'm going to change the Logf to Failf inside cleanupLocalVolume.

framework.Logf("AfterEach: Failed to delete PV and/or PVC: %v", utilerrors.NewAggregate(errs))

to

framework.Failf("Failed to delete PV and/or PVC: %v", utilerrors.NewAggregate(errs))

jeffvance commented on test/e2e/storage/persistent_volumes-local.go:
```diff
@@ -237,9 +485,26 @@ func createLocalPVCPV(config *localTestConfig, volume *localTestVolume) {
+
+func createWriteAndReadCmds(testFileDir string, testFile string, writeTestFileContent string) (string, string) {
+	testFilePath := fmt.Sprintf("%s/%s", testFileDir, testFile)
+	writeCmd := fmt.Sprintf("mkdir -p %s; echo %s > %s; cat %s", testFileDir, writeTestFileContent, testFilePath, testFilePath)
```
`writeCmd` also reads the file. Maybe `writeReadCmd` is a better name. Or maybe it should just write to the file?

ianchakeres replied:
I'll change it to just write to the file to avoid any confusion.

jeffvance commented on test/e2e/storage/persistent_volumes-local.go:
```diff
@@ -237,9 +485,26 @@ func createLocalPVCPV(config *localTestConfig, volume *localTestVolume) {
+	return framework.CreatePod(config.client, config.ns, []*v1.PersistentVolumeClaim{volume.pvc}, false, "")
+}
+
+func createWriteAndReadCmds(testFileDir string, testFile string, writeTestFileContent string) (string, string) {
```
Two reasons to name the return args here: 1) it's slightly more efficient (https://blog.minio.io/golang-internals-part-2-nice-benefits-of-named-return-values-1e95305c8687), 2) it's easier to see which cmd is returned first and which is returned second just read the func sig.

jeffvance commented on test/e2e/storage/persistent_volumes-local.go:
```diff
@@ -105,43 +110,286 @@ var _ = framework.KubeDescribe("[Volume] PersistentVolumes-local [Feature:LocalP
-			podSpec := createLocalPod(config, testVol, readCmd)
-			f.TestContainerOutput("pod consumes PV", podSpec, 0, []string{testContents})
+			By("Creating a pod to read from the PV")
+			_, readCmd := createWriteAndReadCmds(volumeDir, testFile, "" /*writeTestFileContent*/)
```
testFileContent == "" here but the expected output in the next line is `"test-file-content"` so no match should be expected.

ianchakeres replied:
I've added a comment that 

//testFileContent was written during setupLocalVolume

jeffvance replied:
L428 in `runLocalUtil`: there;s a const for the busybox image name in framework.

jeffvance commented on test/e2e/storage/persistent_volumes-local.go:
```diff
@@ -237,9 +485,26 @@ func createLocalPVCPV(config *localTestConfig, volume *localTestVolume) {
+}
+
+func createWriteAndReadCmds(testFileDir string, testFile string, writeTestFileContent string) (string, string) {
+	testFilePath := fmt.Sprintf("%s/%s", testFileDir, testFile)
```
use `path.Join()`

jeffvance commented on test/e2e/storage/persistent_volumes-local.go:
```diff
@@ -389,6 +392,7 @@ func setupLocalVolume(config *localTestConfig, node *v1.Node) *localTestVolume {
 	hostDir := filepath.Join(hostBase, testDirName)
+	//populate volume with testFile containing testFileContent
 	writeCmd, _ := createWriteAndReadCmds(testDir, testFile, testFileContent)
 	framework.Logf("Creating local volume on node %q at path %q", node.Name, hostDir)
```
This could be `By` instead of Logf -- makes the text stand out more. Personal choice.

jeffvance commented:
Seven `It`'s start with:
```
By("Initializing test volume")
testVol = setupLocalVolume(config, node) 
By("Creating local PVC and PV")
createLocalPVCPV(config, testVol)
```
Can this be added to the BeforeEach?

ianchakeres commented:
@jeffvance - I've addressed your comments and pulled the duplicative test initialization into BeforeEach.

ianchakeres commented:
Hi @gnufied & @msau42 - I think this PR is ready to go now. PTAL.

ianchakeres commented:
/test pull-kubernetes-unit

ianchakeres commented:
/test pull-kubernetes-e2e-kops-aws
/test pull-kubernetes-unit

ianchakeres commented:
/test pull-kubernetes-unit
/test pull-kubernetes-e2e-kops-aws

msau42 commented on test/e2e/storage/persistent_volumes-local.go:
```diff
@@ -87,62 +95,257 @@ var _ = framework.KubeDescribe("[Volume] PersistentVolumes-local [Feature:LocalP
-	writeCmd := fmt.Sprintf("mkdir %s; echo %s > %s", testDir, testContents, testFilePath)
-	framework.Logf("Creating local volume on node %q at path %q", node.Name, hostDir)
+	//populate volume with testFile containing testFileContent
+	writeCmd, _ := createWriteAndReadCmds(testDir, testFile, testFileContent)
```
You shouldn't need the second return value.

msau42 commented on test/e2e/storage/persistent_volumes-local.go:
```diff
@@ -87,62 +95,257 @@ var _ = framework.KubeDescribe("[Volume] PersistentVolumes-local [Feature:LocalP
-		It("should be able to mount and write to the volume", func() {
-			By("Initializing test volume")
-			testVol = setupLocalVolume(config, node)
+	Context("one pod with one prebound PVC writes to PV", func() {
```
I think this can fit under the first context since it's also just testing one pod

msau42 commented on test/e2e/storage/persistent_volumes-local.go:
```diff
@@ -87,62 +95,257 @@ var _ = framework.KubeDescribe("[Volume] PersistentVolumes-local [Feature:LocalP
+		})
+	})
+
+	Context("one pod with one prebound PVC reads from PV", func() {
```
Can fit under the first context

msau42 commented on test/e2e/storage/persistent_volumes-local.go:
```diff
@@ -87,62 +95,257 @@ var _ = framework.KubeDescribe("[Volume] PersistentVolumes-local [Feature:LocalP
 		})
 	})
+
+	Context("when two pods request one prebound PVC one after other", func() {
```
Can fit under the 2nd context about two pods accessing a PV serially

ianchakeres commented:
/retest

ianchakeres commented:
This PR https://github.com/kubernetes/kubernetes/pull/48762 should help with docker timeout in e2e-kops-aws.

copejon commented on test/e2e/storage/persistent_volumes-local.go:
```diff
@@ -237,9 +440,41 @@ func createLocalPVCPV(config *localTestConfig, volume *localTestVolume) {
+// Execute a read or write command in a pod
+func podRWCmdExec(pod *v1.Pod, cmd string) string {
+	out, err := podExec(pod, cmd)
+	Expect(err).NotTo(HaveOccurred())
```
Just an observation - in createLocalPod, the error is returned; here the error is handled in scope.  @jeffvance and I have found it's generally better to return errors and handle them at the It() scope.

ianchakeres commented:
@copejon - Thanks for the review. FYI: A few of your comments were on a earlier commit and were already addressed, so you won't see them changed in the next commit.

msau42 commented:
/lgtm

I don't think we need a release note for this

fejta commented:
/test pull-kubernetes-e2e-kops-aws

ianchakeres commented:
@msau42 - I squashed the commits.

msau42 commented:
/lgtm

jeffvance commented:
/test pull-kubernetes-unit

ianchakeres commented:
/assign @gnufied

@gnufied can you please provide ltgm?

gnufied commented on test/e2e/storage/persistent_volumes-local.go:
```diff
@@ -53,96 +54,238 @@ const (
 		var (
 			testVol *localTestVolume
-			node    *v1.Node
 		)
```
very minor nit, while there were 2 variable declared here - old style made sense but with just one variable, it looks out of place

gnufied commented on test/e2e/storage/persistent_volumes-local.go:
```diff
@@ -53,96 +54,238 @@ const (
+
+	Context("when two pods request one prebound PVC one after other", func() {
+		var (
+			testVol *localTestVolume
```
same nit as before. it may just be me. But single variables read better via - `var testVol *localTestVolume`

ianchakeres replied:
I'll fix this nit.

While I rebase due to https://github.com/kubernetes/kubernetes/pull/48781

gnufied reviewed (commented):
Mostly lgtm apart from minor nits.

gnufied commented:
We can fix those nits in follow up commits. Approving for now. 

/approve

gnufied commented:
/approve

msau42 commented:
I created #48832 for this PR.  @ianchakeres please add it to the first comment

ianchakeres commented:
/test pull-kubernetes-federation-e2e-gce

ianchakeres commented:
@gnufied @msau42 - I rebased this PR. Can you please lgtm again?

gnufied commented:
/lgtm

ianchakeres commented:
/test pull-kubernetes-federation-e2e-gce

msau42 commented:
/lgtm

ianchakeres commented:
/test pull-kubernetes-federation-e2e-gce

ianchakeres commented:
/test pull-kubernetes-federation-e2e-gce
